src/transform-sdk/js/main.cc

```diff
@@ -0,0 +1,362 @@
+
+void redpanda_js_source_get_file(char* dst) {
+    std::memcpy(dst, test_source_file.data(), test_source_file.size());
+}
+
+#endif
+}
+
+/**
+ * A custom JS class holding opaque bytes, easily convertable into common JS
+ * types.
+ */
+class record_data {
+public:
+    explicit record_data(bytes_view data)
+      : _data(data) {}
+    record_data(const record_data&) = delete;
+    record_data& operator=(const record_data&) = delete;
+    record_data(record_data&&) = default;
+    record_data& operator=(record_data&&) = default;
+
+    std::expected<qjs::value, qjs::exception>
+    text(JSContext* ctx, std::span<qjs::value> /*params*/) {
+        return qjs::value::string(ctx, std::string_view{_data});
+    }
+
+    std::expected<qjs::value, qjs::exception>
+    json(JSContext* ctx, std::span<qjs::value> /*params*/) {
+        // TODO(rockwood): This is going to be the most common case, this needs
+        // to be zero copy.
+        std::string str;
+        str.append_range(_data);
+        return qjs::value::parse_json(ctx, str);
+    }
+
+    std::expected<qjs::value, qjs::exception>
+    array(JSContext* ctx, std::span<qjs::value> /*params*/) {
+        const std::span data_view = {// NOLINTNEXTLINE(*-const-cast)
+                                     const_cast<uint8_t*>(_data.data()),
+                                     _data.size()};
+        auto array = qjs::value::uint8_array(ctx, data_view);
+        // This memory isn't copied so we need to make sure we
+        // invalid these arrays when the memory is gone.
+        _arrays.push_back(array);
+        return array;
+    }
+
+    ~record_data() {
+        for (qjs::value array : _arrays) {
+            std::ignore = array.detach_uint8_array();
+        }
+    }
+
+    [[nodiscard]] bytes_view data() const { return _data; }
+
+private:
+    bytes_view _data;
+    std::vector<qjs::value> _arrays;
+};
+
+/**
+ * A JS class representation of redpanda::record_writer
+ */
+class record_writer {
+public:
+    explicit record_writer(
+      redpanda::record_writer* writer,
+      qjs::class_factory<record_data>* record_data)
+      : _writer(writer)
+      , _record_data(record_data) {}
+
+    std::expected<qjs::value, qjs::exception>
+    write(JSContext* ctx, std::span<qjs::value> params) {
+        if (params.size() != 1) [[unlikely]] {
+            return std::unexpected(qjs::exception::make(
+              ctx,
+              std::format(
+                "invalid number of parameters to writer.write, got: {}, "
+                "expected: 1",
+                params.size())));
+        }
+        auto& param = params.front();
+        if (!param.is_object()) [[unlikely]] {
+            return std::unexpected(qjs::exception::make(
+              ctx, "expected only object parameters to writer.write"));
+        }
+        auto key = extract_data(ctx, param.get_property("key"));
+        if (!key) [[unlikely]] {
+            return std::unexpected(key.error());
+        }
```

Comment:
(micro)nit: this was slightly confusing to read. looks like "if there's no key, error" at first glance rather than "if extract_data produced an error, propagate the error".

Reply:
Added explicit `has_value` checks.